lib/spack/spack/url_buildcache.py

```diff
@@ -0,0 +1,1129 @@
+        manifest_name: str,
+        manifest: BuildcacheManifest,
+        tmpdir: str,
+        signing_key: Optional[str] = None,
+    ) -> None:
+        # write the manifest to a temporary location
+        manifest_file_name = f"{manifest_name}.manifest.json"
+        manifest_path = os.path.join(tmpdir, manifest_file_name)
+        with open(manifest_path, "w", encoding="utf-8") as f:
+            json.dump(manifest.to_json(), f)
+
+        if signing_key:
+            manifest_path = sign_file(signing_key, manifest_path)
+
+        manifest_destination_url = url_util.join(
+            mirror_url,
+            *cls.get_relative_path_components(BuildcacheComponent.SPECS),
+            manifest_file_name,
+        )
+
+        web_util.push_to_url(manifest_path, manifest_destination_url, keep_original=False)
+
+    @classmethod
+    def push_local_file_as_blob(
+        cls,
+        local_file_path: str,
+        mirror_url: str,
+        manifest_name: str,
+        component_type: BuildcacheComponent,
+        compression: str = "none",
+    ) -> None:
+        cache_class = get_url_buildcache_class()
+        checksum_algo = "sha256"
+        blob_to_push = local_file_path
+        tmpdir = tempfile.mkdtemp()
+
+        try:
+            blob_to_push = os.path.join(tmpdir, os.path.basename(local_file_path))
+            with open(local_file_path, "r", encoding="utf-8") as fd:
+                data_bytes = fd.read().encode("utf-8")
+            data_checksum, data_size = compress_and_write_data(
+                data=data_bytes,
+                output_path=blob_to_push,
+                compression=compression,
+                checksum_algo=checksum_algo,
+            )
+            record = BlobRecord(
+                data_size,
+                cache_class.component_to_content_type(component_type),
+                compression,
+                checksum_algo,
+                data_checksum,
+            )
+            manifest = BuildcacheManifest(
+                layout_version=CURRENT_BUILD_CACHE_LAYOUT_VERSION, data=[record]
+            )
+            cls.push_blob(mirror_url, blob_to_push, record)
+            cls.push_manifest(mirror_url, manifest_name, manifest, tmpdir)
+        finally:
+            shutil.rmtree(tmpdir)
+
+    def push(
+        self,
+        spec: spack.spec.Spec,
+        tarball_path: str,
+        checksum_algorithm: str,
+        tarball_checksum: str,
+        tmpdir: str,
+        signing_key: Optional[str],
+    ) -> None:
+        """Push tarball, specfile, and manifest to the remote mirror
+
+        Pushing should only be done after checking for the pre-existence of a
+        buildcache entry for this spec, and represents a force push if one is
+        found.  Thus, any pre-existing files are first removed.
+        """
+
+        spec_dict = spec.to_dict(hash=ht.dag_hash)
+        layout_version = self.get_layout_version()
+        spec_dict["buildcache_layout_version"] = layout_version
```

Comment:
Is this relevant now?

Reply:
This is still used when downloading oci image tarballs to check validity of the spec.  If the layout version in the spec file is too new, oci download tarball ignores it.

Reply:
I think the logic to make drop this now is a bit messy. Maybe in a future version of Spack where `v2` build cache support is dropped this can be revisited.